File: langs/csharp/src/VMx/Forwarding/ForwardingComponentVM.cs
#pragma warning disable CA1051 // Protected field _wrapped is intentional: subclasses use it to override individual members
using System.ComponentModel;
using System.Windows.Input;
using VMx.Components;
using VMx.Services;

namespace VMx.Forwarding;

/// <summary>
/// Abstract forwarding decorator for <see cref="IComponentVM{M}"/>.
/// Every public member delegates to <see cref="_wrapped"/> by default.
/// Subclasses override individual members to customize behaviour.
///
/// See spec/09-forwarding.md §ForwardingComponentVM and FWD-001/FWD-002 in spec/12-conformance.md.
/// </summary>
/// <typeparam name="M">The model type.</typeparam>
public abstract class ForwardingComponentVM<M> : IComponentVM<M>, IComponentVMInternals, IDisposable
{
    /// <summary>The wrapped instance that all members delegate to by default.</summary>
    protected readonly IComponentVM<M> _wrapped;
    private IParentCompositeVM? _parent;

    private sealed class ParentAdapter(
        IParentCompositeVM parent,
        ForwardingComponentVM<M> wrapper) : IParentCompositeVM
    {
        internal ForwardingComponentVM<M> RetainedWrapper => wrapper;
        public IComponentVM? Owner => parent.Owner;
        public IParentCompositeVM? OwnerParent => parent.OwnerParent;
        public bool SupportsChildSelection => parent.SupportsChildSelection;
        public IComponentVM? CurrentChild =>
            ReferenceEquals(parent.CurrentChild, wrapper) ? wrapper._wrapped : parent.CurrentChild;
        public void SelectChild(IComponentVM vm) => parent.SelectChild(wrapper);
        public void DeselectChild(IComponentVM vm) => parent.DeselectChild(wrapper);
        public bool ContainsChild(IComponentVM vm) => parent.ContainsChild(wrapper);
        public ParentTransferToken DetachForTransfer(IComponentVM vm) =>
            parent.DetachForTransfer(wrapper);
    }

    private sealed class WrappedParentAdapter(
        IParentCompositeVM parent,
        ForwardingComponentVM<M> wrapper,
        IComponentVM wrapped) : IParentCompositeVM
    {
        public IComponentVM? Owner => parent.Owner;
        public IParentCompositeVM? OwnerParent => parent.OwnerParent;
        public bool SupportsChildSelection => parent.SupportsChildSelection;
        public IComponentVM? CurrentChild =>
            ReferenceEquals(parent.CurrentChild, wrapped) ? wrapper : parent.CurrentChild;
        public void SelectChild(IComponentVM vm) => parent.SelectChild(wrapped);
        public void DeselectChild(IComponentVM vm) => parent.DeselectChild(wrapped);
        public bool ContainsChild(IComponentVM vm) => parent.ContainsChild(wrapped);
        public ParentTransferToken DetachForTransfer(IComponentVM vm)
        {
            var retainedWrapper = (parent as ParentAdapter)?.RetainedWrapper;
            var staged = parent.DetachForTransfer(wrapped);
            return new ParentTransferToken(
                commit: () =>
                {
                    try { staged.Commit(); }
                    finally
                    {
                        if (retainedWrapper is not null &&
                            !ReferenceEquals(retainedWrapper, wrapper))
                            retainedWrapper.ClearDirectParent();
                    }
                },
                rollback: staged.Rollback);
        }
    }

    /// <summary>Initialises the decorator with the instance to wrap.</summary>
    protected ForwardingComponentVM(IComponentVM<M> wrapped)
        => _wrapped = wrapped ?? throw new ArgumentNullException(nameof(wrapped));

    private void ClearDirectParent() => _parent = null;

    // ── IComponentVM: identity ─────────────────────────────────────────────

    /// <inheritdoc/>
    public virtual string Name => _wrapped.Name;

    /// <inheritdoc/>
    public virtual string Hint => _wrapped.Hint;

    /// <inheritdoc/>
    public virtual ViewModelType Type => _wrapped.Type;

    // ── IComponentVM: state ────────────────────────────────────────────────

    /// <inheritdoc/>
    public virtual bool IsCurrent => _wrapped.IsCurrent;

    /// <inheritdoc/>
    public virtual bool IsConstructed => _wrapped.IsConstructed;

    /// <inheritdoc/>
    public virtual Lifecycle.ConstructionStatus Status => _wrapped.Status;

    /// <inheritdoc/>
    public virtual IMessageHub Hub => _wrapped.Hub;

    // ── IComponentVM<M>: model ─────────────────────────────────────────────

    /// <inheritdoc/>
    public virtual M Model
    {
        get => _wrapped.Model;
        set => _wrapped.Model = value;
    }

    /// <inheritdoc/>
    public virtual string ModeledHint => _wrapped.ModeledHint;

    /// <inheritdoc/>
    public virtual void RepublishModel() => _wrapped.RepublishModel();

    // ── IComponentVM: commands ─────────────────────────────────────────────

    /// <inheritdoc/>
    public virtual ICommand SelectCommand => _wrapped.SelectCommand;

    /// <inheritdoc/>
    public virtual ICommand DeselectCommand => _wrapped.DeselectCommand;

    /// <inheritdoc/>
    public virtual ICommand SelectNextCommand => _wrapped.SelectNextCommand;

    /// <inheritdoc/>
    public virtual ICommand SelectPreviousCommand => _wrapped.SelectPreviousCommand;

    /// <inheritdoc/>
    public virtual ICommand ReconstructCommand => _wrapped.ReconstructCommand;

    // ── IComponentVM: lifecycle ────────────────────────────────────────────

    /// <inheritdoc/>
    public virtual bool CanConstruct() => _wrapped.CanConstruct();

    /// <inheritdoc/>
    public virtual void Construct() => _wrapped.Construct();

    /// <inheritdoc/>
    public virtual Task ConstructAsync() => _wrapped.ConstructAsync();

    /// <inheritdoc/>
    public virtual bool CanDestruct() => _wrapped.CanDestruct();

    /// <inheritdoc/>
    public virtual void Destruct() => _wrapped.Destruct();

    /// <inheritdoc/>
    public virtual Task DestructAsync() => _wrapped.DestructAsync();

    /// <inheritdoc/>
    public virtual bool CanReconstruct() => _wrapped.CanReconstruct();

    /// <inheritdoc/>
    public virtual void Reconstruct() => _wrapped.Reconstruct();

    /// <inheritdoc/>
    public virtual Task ReconstructAsync() => _wrapped.ReconstructAsync();

    // ── IComponentVM: selection ────────────────────────────────────────────

    /// <inheritdoc/>
    public virtual bool CanSelect() => _wrapped.CanSelect();

    /// <inheritdoc/>
    public virtual void Select() => _wrapped.Select();

    /// <inheritdoc/>
    public virtual bool CanDeselect() => _wrapped.CanDeselect();

    /// <inheritdoc/>
    public virtual void Deselect() => _wrapped.Deselect();

    // ── INotifyPropertyChanged ─────────────────────────────────────────────

    /// <inheritdoc/>
    public event PropertyChangedEventHandler? PropertyChanged
    {
        add => _wrapped.PropertyChanged += value;
        remove => _wrapped.PropertyChanged -= value;
    }

    // ── IDisposable ────────────────────────────────────────────────────────

    /// <inheritdoc/>
    public virtual void Dispose()
    {
        _wrapped.Dispose();
    }

    IParentCompositeVM? IComponentVMInternals.Parent
    {
        get
        {
            if (_parent is not null) return _parent;
            var wrappedParent = _wrapped.GetParent();
            return wrappedParent is null
                ? null
                : new WrappedParentAdapter(wrappedParent, this, _wrapped);
        }
    }

    void IComponentVMInternals.SetParent(IParentCompositeVM? parent)
    {
        _parent = parent;
        _wrapped.SetParent(parent is null ? null : new ParentAdapter(parent, this));
    }

    void IComponentVMInternals.SetIsCurrent(bool value) => _wrapped.SetIsCurrent(value);

    bool IComponentVMInternals.CommitIsCurrent(bool value) => _wrapped.CommitIsCurrent(value);

    void IComponentVMInternals.PublishIsCurrent() => _wrapped.PublishIsCurrent();

    Task IComponentVMInternals.ConstructOrJoinAsync() => _wrapped.ConstructOrJoinAsync();
}
#pragma warning restore CA1051

File: langs/typescript/src/forwarding/forwardingComponentVM.ts
/**
 * ForwardingComponentVM<M> — forwarding decorator for IComponentVMOf<M>.
 *
 * Every member delegates to the wrapped VM by default.
 * Subclasses override individual members to customize behavior.
 *
 * See spec/09-forwarding.md §ForwardingComponentVM.
 */
import type { Observable } from "rxjs";
import type { ConstructionStatus } from "../lifecycle/status.js";
import type { IComponentVMOf, ViewModelType } from "../components/types.js";
import { RelayCommand } from "../commands/relayCommand.js";
import type { IMessageHub } from "../services/messageHub.js";
import {
  ComponentVMBase,
  type IOwningParentVM,
  ParentTransfer,
} from "../components/componentVMBase.js";
import { NullDispatcher } from "../services/nullDispatcher.js";

class ForwardingParent<M> implements IOwningParentVM {
  constructor(
    private readonly parent: IOwningParentVM,
    private readonly wrapper: ForwardingComponentVM<M>,
  ) {}

  get owner(): ComponentVMBase { return this.parent.owner; }
  get ownerParent(): IOwningParentVM | null { return this.parent.ownerParent; }
  get supportsChildSelection(): boolean { return this.parent.supportsChildSelection; }
  get currentChild(): ComponentVMBase | null {
    return this.parent.currentChild === this.wrapper
      ? this.wrapper.wrappedBase
      : this.parent.currentChild;
  }
  selectChild(_vm: ComponentVMBase): void { this.parent.selectChild(this.wrapper); }
  deselectChild(_vm: ComponentVMBase): void { this.parent.deselectChild(this.wrapper); }
  containsChild(_vm: ComponentVMBase): boolean { return this.parent.containsChild(this.wrapper); }
  detachForTransfer(_vm: ComponentVMBase): ParentTransfer {
    return this.parent.detachForTransfer(this.wrapper);
  }
}

export class ForwardingComponentVM<M> extends ComponentVMBase implements IComponentVMOf<M> {
  protected readonly _wrapped: IComponentVMOf<M>;

  constructor(wrapped: IComponentVMOf<M>) {
    super({
      name: wrapped.name,
      hint: wrapped.hint,
      hub: wrapped.hub,
      dispatcher: NullDispatcher.INSTANCE,
    });
    this._wrapped = wrapped;
  }

  get wrappedBase(): ComponentVMBase {
    if (!(this._wrapped instanceof ComponentVMBase)) {
      throw new TypeError("A forwarding container child must wrap a VMx ComponentVMBase");
    }
    return this._wrapped;
  }

  override get _parent(): IOwningParentVM | null { return super._parent; }
  override set _parent(parent: IOwningParentVM | null) {
    super._parent = parent;
    if (this._wrapped instanceof ComponentVMBase) {
      this._wrapped._parent = parent === null ? null : new ForwardingParent(parent, this);
    }
  }

  get name(): string { return this._wrapped.name; }
  get hint(): string { return this._wrapped.hint; }
  get type(): ViewModelType { return this._wrapped.type; }
  get isCurrent(): boolean { return this._wrapped.isCurrent; }
  get isConstructed(): boolean { return this._wrapped.isConstructed; }
  get status(): ConstructionStatus { return this._wrapped.status; }
  get hub(): IMessageHub { return this._wrapped.hub; }
  get model(): M { return this._wrapped.model; }
  // Delegate the model setter too (spec/09-forwarding.md §1 — the decorator
  // forwards every member; the modeled component's `model` is settable). Matches
  // C#/Python/Swift ForwardingComponentVM and TS ForwardingCompositeVM.set current.
  set model(value: M) { this._wrapped.model = value; }
  get modeledHint(): string { return this._wrapped.modeledHint; }
  republishModel(): void { this._wrapped.republishModel(); }
  get propertyChanged(): Observable<string> { return this._wrapped.propertyChanged; }
  get selectCommand(): RelayCommand { return this._wrapped.selectCommand as RelayCommand; }
  get deselectCommand(): RelayCommand { return this._wrapped.deselectCommand as RelayCommand; }
  get selectNextCommand(): RelayCommand { return this._wrapped.selectNextCommand as RelayCommand; }
  get selectPreviousCommand(): RelayCommand {
    return this._wrapped.selectPreviousCommand as RelayCommand;
  }
  get reconstructCommand(): RelayCommand { return this._wrapped.reconstructCommand as RelayCommand; }

  override _setIsCurrent(value: boolean): void {
    if (this._wrapped instanceof ComponentVMBase) this._wrapped._setIsCurrent(value);
  }

  canConstruct(): boolean { return this._wrapped.canConstruct(); }
  construct(): void { this._wrapped.construct(); }
  canDestruct(): boolean { return this._wrapped.canDestruct(); }
  destruct(): void { this._wrapped.destruct(); }
  canReconstruct(): boolean { return this._wrapped.canReconstruct(); }
  reconstruct(): void { this._wrapped.reconstruct(); }
  dispose(): void { this._wrapped.dispose(); }
  canSelect(): boolean { return this._wrapped.canSelect(); }
  select(): void { this._wrapped.select(); }
  canDeselect(): boolean { return this._wrapped.canDeselect(); }
  deselect(): void { this._wrapped.deselect(); }
}
